File: data_structure/prefix_sum/238_product_of_array_except_self.cc
#include <vector>

// 2024/01/05: 直接看答案

std::vector<int> productExceptSelf(std::vector<int>& nums) {
  // 关键思路: 左右乘积列表
  int n = nums.size();
  std::vector<int> left(n, 0);
  std::vector<int> right(n, 0);
  left[0] = 1;
  for (int i = 1; i < n; ++i) {
    left[i] = left[i - 1] * nums[i - 1];
  }
  right[n - 1] = 1;
  for (int i = n - 2; i >= 0; --i) {
    right[i] = right[i + 1] * nums[i + 1];
  }
  std::vector<int> res(n, 0);
  for (int i = 0; i < n; ++i) {
    res[i] = left[i] * right[i];
  }
  return res;
}

File: data_structure/prefix_sum/238_product_of_array_except_self.py
from typing import List


# 2026/05/11: 自己写出答案, 时间复杂度和空间复杂度都是 O(n).
def product_except_self(nums: List[int]) -> List[int]:
    left_products, right_products, res = (
        [0] * len(nums),
        [0] * len(nums),
        [0] * len(nums),
    )

    left_products[0] = 1
    for i in range(1, len(nums)):
        left_products[i] = left_products[i - 1] * nums[i - 1]

    right_products[-1] = 1
    for i in range(len(nums) - 2, -1, -1):
        right_products[i] = right_products[i + 1] * nums[i + 1]

    for i in range(0, len(nums)):
        res[i] = left_products[i] * right_products[i]

    return res


nums1 = [1, 2, 3, 4]
nums2 = [-1, 1, 0, -3, 3]
print(product_except_self(nums1))
print(product_except_self(nums2))
